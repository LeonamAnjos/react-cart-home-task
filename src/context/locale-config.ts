export type LocaleType = "de-DE" | "en-US";

type LocaleConfig = {
  currency: string;
};

export const localeMap: Record<LocaleType, LocaleConfig> = {
  "de-DE": { currency: "EUR" },
  "en-US": { currency: "USD" },
};
