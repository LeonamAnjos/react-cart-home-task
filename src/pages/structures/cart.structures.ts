import { Product } from "./product.structures";

export type CartItem = {
  product: Product;
  coverage: number;
  price: number;
};
