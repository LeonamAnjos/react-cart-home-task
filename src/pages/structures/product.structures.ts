export type Product = {
  id: number;
  name: string;
  minCoverage: number;
  maxCoverage: number;
  risk: number;
};

export const calcPrice = (coverage: number, risk: number): number =>
  coverage * risk;
