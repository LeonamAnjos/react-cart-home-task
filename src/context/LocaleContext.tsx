import React, { createContext, ReactNode, useMemo, useState } from "react";
import { localeMap, LocaleType } from "./locale-config";

type LocaleState = {
  getLocale: () => LocaleType | undefined;
  setLocale: (locale: LocaleType) => void;
  currencyFormat: (num: number) => string;
};

const LocaleContext = createContext<LocaleState>({
  getLocale: (): LocaleType | undefined => undefined,
  setLocale: (locale: LocaleType): void => {
    /* EMPTY */
  },
  currencyFormat: (num: number): string => "",
});

export interface LocaleProviderProps {
  children?: ReactNode;
  locale: LocaleType;
}

export const LocaleProvider = ({ children, locale }: LocaleProviderProps) => {
  const [value, setValue] = useState(locale);

  const contextValue = useMemo(() => {
    const formatter = new Intl.NumberFormat(value, {
      style: "currency",
      currency: localeMap[value]?.currency,
      minimumFractionDigits: 2,
    });

    return {
      getLocale: () => value,
      setLocale: (locale: LocaleType) => setValue(locale),
      currencyFormat: (num: number): string => formatter.format(num),
    };
  }, [value, setValue]);

  return (
    <LocaleContext.Provider value={contextValue}>
      {children}
    </LocaleContext.Provider>
  );
};

export default LocaleContext;
