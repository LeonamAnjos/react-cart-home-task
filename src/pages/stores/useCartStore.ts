import create from "zustand";
import { persist } from "zustand/middleware";
import { CartItem } from "../structures/cart.structures";

type CartState = {
  items: CartItem[];
  addItem: (item: CartItem) => void;
  removeItem: (item: CartItem) => void;
};

export const useCartStore = create<CartState, [["zustand/persist", CartState]]>(
  persist(
    (set) => ({
      items: [] as CartItem[],
      addItem: (item: CartItem) => {
        set((state: CartState) => ({
          ...state,
          items: [
            ...state.items.filter(
              ({ product }) => product.id !== item.product.id
            ),
            item,
          ],
        }));
      },
      removeItem: (item: CartItem) => {
        set((state: CartState) => ({
          ...state,
          items: [
            ...state.items.filter(
              ({ product }) => product.id !== item.product.id
            ),
          ],
        }));
      },
    }),
    {
      name: "cart-store",
    }
  )
);
