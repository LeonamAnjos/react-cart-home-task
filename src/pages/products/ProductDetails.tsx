import React, { useContext, useState } from "react";
import Avatar from "@mui/material/Avatar";
import { calcPrice, Product } from "../structures/product.structures";
import ElementCard from "../../components/ElementCard";
import LocaleContext from "../../context/LocaleContext";
import CoverageSlider from "../commons/CoverageSlider";
import AddToCartButton from "./AddToCartButton";
import { useCartStore } from "../stores/useCartStore";
import { firstLetter } from "../utils/utils";
import CoveragePriceCaption from "../commons/CoveragePriceCaption";

type ProductDetailsProps = {
  product: Product;
};

const formatSubTitle = (minCoverage: string, maxCoverage: string) =>
  `Coverage from ${minCoverage} to ${maxCoverage}`;

const ProductDetails = ({ product }: ProductDetailsProps) => {
  const [coverage, setCoverage] = useState(product.maxCoverage);
  const [price, setPrice] = useState(calcPrice(coverage, product.risk));

  const { addItem } = useCartStore();
  const { currencyFormat } = useContext(LocaleContext);

  const handleCoverageChange = (coverage: number): void => {
    setCoverage(coverage);
    setPrice(calcPrice(coverage, product.risk));
  };

  const handleAddToCartClick = (): void => {
    addItem({ product, coverage, price });
  };

  return (
    <ElementCard
      avatar={<Avatar>{firstLetter(product.name)}</Avatar>}
      title={product.name}
      subTitle={formatSubTitle(
        currencyFormat(product.minCoverage),
        currencyFormat(product.maxCoverage)
      )}
      actions={<AddToCartButton onClick={handleAddToCartClick} />}
    >
      <CoverageSlider
        minCoverage={product.minCoverage}
        maxCoverage={product.maxCoverage}
        onCoverageChange={handleCoverageChange}
      />

      <CoveragePriceCaption coverage={coverage} price={price} />
    </ElementCard>
  );
};

export default ProductDetails;
