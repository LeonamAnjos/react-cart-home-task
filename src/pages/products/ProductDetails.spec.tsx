import { render, screen, fireEvent } from "@testing-library/react";
import { LocaleProvider } from "../../context/LocaleContext";
import { Product } from "../structures/product.structures";
import { useCartStore } from "../stores/useCartStore";
import ProductDetails from "./ProductDetails";

const originalState = useCartStore.getState();
beforeEach(() => {
  useCartStore.setState(originalState);
});

describe("<ProductDetails />", () => {
  const customRender = (product?: Product) => {
    const customProduct = {
      id: 1,
      name: "Test Product",
      minCoverage: 10,
      maxCoverage: 90,
      risk: 0.3,
      ...product,
    };

    return render(
      <LocaleProvider locale="en-US">
        <ProductDetails product={customProduct} />
      </LocaleProvider>
    );
  };

  it("SHOULD render", () => {
    const { container } = customRender();

    expect(container).toMatchSnapshot();
  });

  it("SHOULD add product to the cart", () => {
    customRender();

    const element = screen.getByTestId("add-to-cart-button");
    fireEvent.click(element);

    const { items } = useCartStore.getState();

    expect(items.length).toBe(1);
    expect(items[0].coverage).toBe(90);
    expect(items[0].price).toBe(27);
    expect(items[0].product.name).toBe("Test Product");
  });

  it("SHOULD change the coverage and price", () => {
    const { container } = customRender();

    const element = screen.getByTestId("coverage-slider");
    fireEvent.mouseDown(element, { clientX: -1 });

    expect(container).toMatchSnapshot();
  });
});
