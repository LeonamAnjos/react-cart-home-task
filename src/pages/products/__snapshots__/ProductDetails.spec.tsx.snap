// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductDetails /> SHOULD change the coverage and price 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-16cqd0p-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
        >
          T
        </div>
      </div>
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
        >
          Test Product
        </span>
        <span
          class="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-83ijpv-MuiTypography-root"
        >
          Coverage from $10.00 to $90.00
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <span
        class="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root css-187mznn-MuiSlider-root"
        data-testid="coverage-slider"
      >
        <span
          class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
        />
        <span
          class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
          style="left: 0%; width: 0%;"
        />
        <span
          class="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb Mui-active css-eg0mwd-MuiSlider-thumb"
          data-index="0"
          style="left: 0%;"
        >
          <input
            aria-orientation="horizontal"
            aria-valuemax="90"
            aria-valuemin="10"
            aria-valuenow="10"
            data-index="0"
            max="90"
            min="10"
            step="10"
            style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
            type="range"
            value="10"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-valueLabelOpen MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
          >
            <span
              class="MuiSlider-valueLabelCircle"
            >
              <span
                class="MuiSlider-valueLabelLabel"
              >
                10
              </span>
            </span>
          </span>
        </span>
      </span>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-bjfanr-MuiTypography-root"
          >
            Coverage: $10.00
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-bjfanr-MuiTypography-root"
          >
            Price: $3.00
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiCardActions-root css-1rwjz6-MuiCardActions-root"
    >
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
        data-testid="add-to-cart-button"
        style="margin-left: auto;"
        tabindex="0"
        type="button"
      >
        Add
        <span
          class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddShoppingCartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-9.83-3.25.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`<ProductDetails /> SHOULD render 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-16cqd0p-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-2s90m6-MuiAvatar-root"
        >
          T
        </div>
      </div>
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
        >
          Test Product
        </span>
        <span
          class="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-83ijpv-MuiTypography-root"
        >
          Coverage from $10.00 to $90.00
        </span>
      </div>
    </div>
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <span
        class="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root css-187mznn-MuiSlider-root"
        data-testid="coverage-slider"
      >
        <span
          class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
        />
        <span
          class="MuiSlider-track css-1gv0vcd-MuiSlider-track"
          style="left: 0%; width: 100%;"
        />
        <span
          class="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
          data-index="0"
          style="left: 100%;"
        >
          <input
            aria-orientation="horizontal"
            aria-valuemax="90"
            aria-valuemin="10"
            aria-valuenow="90"
            data-index="0"
            max="90"
            min="10"
            step="10"
            style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
            type="range"
            value="90"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
          >
            <span
              class="MuiSlider-valueLabelCircle"
            >
              <span
                class="MuiSlider-valueLabelLabel"
              >
                90
              </span>
            </span>
          </span>
        </span>
      </span>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-bjfanr-MuiTypography-root"
          >
            Coverage: $90.00
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-bjfanr-MuiTypography-root"
          >
            Price: $27.00
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiCardActions-root css-1rwjz6-MuiCardActions-root"
    >
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
        data-testid="add-to-cart-button"
        style="margin-left: auto;"
        tabindex="0"
        type="button"
      >
        Add
        <span
          class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddShoppingCartIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-9.83-3.25.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4h-.01l-1.1 2-2.76 5H8.53l-.13-.27L6.16 6l-.95-2-.94-2H1v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.13 0-.25-.11-.25-.25z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
